File: 27-Interpreter/src/go/not_expression.go
package main

// NotExpression is a Non-terminal Expression — logical NOT of one sub-expression.
type NotExpression struct {
	expression Expression
}

func NewNotExpression(expression Expression) *NotExpression {
	return &NotExpression{expression: expression}
}

func (n *NotExpression) Interpret(context *Context) bool {
	return !n.expression.Interpret(context)
}

File: 27-Interpreter/src/rust/not_expression.rs
use crate::expression::Expression;
use crate::context::Context;

/// Non-terminal Expression — logical NOT of one sub-expression.
pub struct NotExpression {
    expression: Box<dyn Expression>,
}

impl NotExpression {
    pub fn new(expression: Box<dyn Expression>) -> Self {
        NotExpression { expression }
    }
}

impl Expression for NotExpression {
    fn interpret(&self, context: &Context) -> bool {
        !self.expression.interpret(context)
    }
}
